course/api/views.py

```diff
@@ -64,6 +95,112 @@ def get_queryset(self):
     def get_object(self):
         return self.get_member_object('instance', 'Course')
 
+    def set_teachers(self, course: CourseInstance, data: dict) -> None:
+        teachers = []
+        for i in data['teachers']:
+            userprofile: UserProfile = None
+            try:
+                user = User.objects.get(username=i)
+            except User.DoesNotExist:
+                # If user does not exist, create a new user.
+                # If external authentication (e.g. Shibboleth) is used, other
+                # attributes will be updated when user logs in for the first time.
+                user = User.objects.create_user(i)
+
+            userprofile = UserProfile.objects.get(user=user)
+            teachers.append(userprofile)
+
+        course.set_teachers(teachers)
+
+    def create(self, request: HttpRequest, version) -> Response:
+        if (not request.user.is_superuser):
+            return Response(
+                'Only admins can add new course instance.',
+                status=status.HTTP_403_FORBIDDEN
+            )
+
+        data = request.data
+        form = CourseInstanceForm(data)
+
+        if not form.is_valid():
+            return Response({'status': 'error', 'errors': form.errors}, status=status.HTTP_400_BAD_REQUEST)
+
+        else:
+            code = form.cleaned_data.get('code')
+            course: Course = Course.objects.filter(code=code).first()
```

Comment:
Is it safer to fail fast (without modifying anything) if the course code is not unique? In the worst case, people modify the wrong Course when they don't realize they have duplicate course codes in the database.